File: pos/ios/Xcode/MisePOS/MisePOSUITests/MisePOS/Network/Protocol.swift
import Foundation

// MARK: - Client → Server

enum ClientMessage: Encodable {
    case auth(deviceId: String, pin: String, role: String)
    case ping
    case tableOpen(requestId: String, tableId: String, guests: Int)
    case tableClose(requestId: String, tableId: String)
    case orderItemAdd(requestId: String, orderId: String, item: OrderItemInput)
    case orderItemRemove(requestId: String, orderId: String, itemId: String)
    case orderSendToKitchen(requestId: String, orderId: String, course: Int?)
    case orderApplyDiscount(requestId: String, orderId: String, amount: Double)
    case orderPay(requestId: String, orderId: String, payment: PaymentInput)
    case kitchenItemReady(requestId: String, itemId: String)
    case kitchenTicketDone(requestId: String, orderId: String)
    case menuToggleItem(requestId: String, itemId: String, available: Bool)
    case syncRequest

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: AnyCodingKey.self)
        switch self {
        case .auth(let did, let pin, let role):
            try c.encode("auth", forKey: .key("type"))
            try c.encode(did, forKey: .key("device_id"))
            try c.encode(pin, forKey: .key("pin"))
            try c.encode(role, forKey: .key("role"))
        case .ping:
            try c.encode("ping", forKey: .key("type"))
        case .tableOpen(let rid, let tid, let guests):
            try c.encode("table.open", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(tid, forKey: .key("table_id"))
            try c.encode(guests, forKey: .key("guests"))
        case .tableClose(let rid, let tid):
            try c.encode("table.close", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(tid, forKey: .key("table_id"))
        case .orderItemAdd(let rid, let oid, let item):
            try c.encode("order.item.add", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(oid, forKey: .key("order_id"))
            try c.encode(item, forKey: .key("item"))
        case .orderItemRemove(let rid, let oid, let iid):
            try c.encode("order.item.remove", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(oid, forKey: .key("order_id"))
            try c.encode(iid, forKey: .key("item_id"))
        case .orderSendToKitchen(let rid, let oid, let course):
            try c.encode("order.send_to_kitchen", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(oid, forKey: .key("order_id"))
            if let course { try c.encode(course, forKey: .key("course")) }
        case .orderApplyDiscount(let rid, let oid, let amount):
            try c.encode("order.apply_discount", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(oid, forKey: .key("order_id"))
            try c.encode(amount, forKey: .key("amount"))
        case .orderPay(let rid, let oid, let payment):
            try c.encode("order.pay", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(oid, forKey: .key("order_id"))
            try c.encode(payment, forKey: .key("payment"))
        case .kitchenItemReady(let rid, let iid):
            try c.encode("kitchen.item.ready", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(iid, forKey: .key("item_id"))
        case .kitchenTicketDone(let rid, let oid):
            try c.encode("kitchen.ticket.done", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(oid, forKey: .key("order_id"))
        case .menuToggleItem(let rid, let iid, let avail):
            try c.encode("menu.toggle_item", forKey: .key("type"))
            try c.encode(rid, forKey: .key("request_id"))
            try c.encode(iid, forKey: .key("item_id"))
            try c.encode(avail, forKey: .key("available"))
        case .syncRequest:
            try c.encode("sync.request", forKey: .key("type"))
        }
    }
}

// MARK: - Server → Client

enum ServerMessage: Decodable {
    case authOk(deviceId: String, role: String, snapshot: Snapshot)
    case authError(message: String)
    case pong
    case ack(requestId: String)
    case error(requestId: String?, message: String)
    case stateTables([TableState])
    case stateOrder(OrderState)
    case stateMenuItem(itemId: String, isAvailable: Bool)
    case stateKdsItem(itemId: String, status: String, orderId: String)
    case notifyItemReady(orderId: String, itemName: String, tableNumber: String)
    case notifyTicketDone(orderId: String, tableNumber: String)
    case paymentOk(orderId: String, paymentId: String, change: Double)
    case stateSession(POSSession)
    case unknown

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: AnyCodingKey.self)
        let type = try c.decode(String.self, forKey: .key("type"))
        switch type {
        case "auth.ok":
            self = .authOk(
                deviceId: try c.decode(String.self, forKey: .key("device_id")),
                role: try c.decode(String.self, forKey: .key("role")),
                snapshot: try c.decode(Snapshot.self, forKey: .key("snapshot"))
            )
        case "auth.error":
            self = .authError(message: try c.decode(String.self, forKey: .key("message")))
        case "pong":
            self = .pong
        case "ack":
            self = .ack(requestId: try c.decode(String.self, forKey: .key("request_id")))
        case "error":
            self = .error(
                requestId: try? c.decode(String.self, forKey: .key("request_id")),
                message: try c.decode(String.self, forKey: .key("message"))
            )
        case "state.tables":
            self = .stateTables(try c.decode([TableState].self, forKey: .key("tables")))
        case "state.order":
            self = .stateOrder(try c.decode(OrderState.self, forKey: .key("order")))
        case "state.menu_item":
            self = .stateMenuItem(
                itemId: try c.decode(String.self, forKey: .key("item_id")),
                isAvailable: try c.decode(Bool.self, forKey: .key("is_available"))
            )
        case "state.kds_item":
            self = .stateKdsItem(
                itemId: try c.decode(String.self, forKey: .key("item_id")),
                status: try c.decode(String.self, forKey: .key("status")),
                orderId: try c.decode(String.self, forKey: .key("order_id"))
            )
        case "notify.item_ready":
            self = .notifyItemReady(
                orderId: try c.decode(String.self, forKey: .key("order_id")),
                itemName: try c.decode(String.self, forKey: .key("item_name")),
                tableNumber: try c.decode(String.self, forKey: .key("table_number"))
            )
        case "notify.ticket_done":
            self = .notifyTicketDone(
                orderId: try c.decode(String.self, forKey: .key("order_id")),
                tableNumber: try c.decode(String.self, forKey: .key("table_number"))
            )
        case "payment.ok":
            self = .paymentOk(
                orderId: try c.decode(String.self, forKey: .key("order_id")),
                paymentId: try c.decode(String.self, forKey: .key("payment_id")),
                change: try c.decode(Double.self, forKey: .key("change"))
            )
        case "state.session":
            self = .stateSession(try c.decode(POSSession.self, forKey: .key("session")))
        default:
            self = .unknown
        }
    }
}

// MARK: - Input types

struct OrderItemInput: Encodable {
    let id: String
    let menuItemId: String
    let menuItemName: String
    let qty: Int
    let unitPrice: Double
    let modifiers: [ModifierSelection]
    let course: Int
    let note: String?

    enum CodingKeys: String, CodingKey {
        case id, qty, modifiers, course, note
        case menuItemId = "menu_item_id"
        case menuItemName = "menu_item_name"
        case unitPrice = "unit_price"
    }
}

struct ModifierSelection: Codable {
    let groupId: String
    let groupName: String
    let optionId: String
    let optionName: String
    let priceDelta: Double

    enum CodingKeys: String, CodingKey {
        case groupId = "group_id"
        case groupName = "group_name"
        case optionId = "option_id"
        case optionName = "option_name"
        case priceDelta = "price_delta"
    }
}

struct PaymentInput: Encodable {
    let id: String
    let method: String
    let amount: Double
    let tipAmount: Double
    let changeAmount: Double
    let cashTendered: Double?
    let stripePaymentIntentId: String?

    enum CodingKeys: String, CodingKey {
        case id, method, amount
        case tipAmount = "tip_amount"
        case changeAmount = "change_amount"
        case cashTendered = "cash_tendered"
        case stripePaymentIntentId = "stripe_payment_intent_id"
    }
}

// MARK: - Coding key helper

struct AnyCodingKey: CodingKey {
    var stringValue: String
    var intValue: Int?
    init(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { self.intValue = intValue; self.stringValue = "\(intValue)" }
    static func key(_ s: String) -> AnyCodingKey { .init(stringValue: s) }
}

File: pos/server/src/ws/protocol.ts
// ── Client → Server ────────────────────────────────────────────────────────

export type ClientMessage =
  | { type: "auth"; device_id: string; pin: string; role: string }
  | { type: "ping" }
  | { type: "table.open"; request_id: string; table_id: string; guests: number }
  | { type: "table.close"; request_id: string; table_id: string }
  | { type: "order.item.add"; request_id: string; order_id: string; item: OrderItemInput }
  | { type: "order.item.remove"; request_id: string; order_id: string; item_id: string }
  | { type: "order.send_to_kitchen"; request_id: string; order_id: string; course?: number }
  | { type: "order.apply_discount"; request_id: string; order_id: string; amount: number }
  | { type: "order.pay"; request_id: string; order_id: string; payment: PaymentInput }
  | { type: "kitchen.item.ready"; request_id: string; item_id: string }
  | { type: "kitchen.ticket.done"; request_id: string; order_id: string }
  | { type: "menu.toggle_item"; request_id: string; item_id: string; available: boolean }
  | { type: "sync.request" }

export interface OrderItemInput {
  id: string
  menu_item_id: string
  menu_item_name: string
  qty: number
  unit_price: number
  modifiers: ModifierSelection[]
  course: number
  note?: string
}

export interface ModifierSelection {
  group_id: string
  group_name: string
  option_id: string
  option_name: string
  price_delta: number
}

export interface PaymentInput {
  id: string
  method: "cash" | "card" | "mixed" | "voucher"
  amount: number
  tip_amount: number
  change_amount: number
  cash_tendered?: number
  stripe_payment_intent_id?: string
}

// ── Server → Client ────────────────────────────────────────────────────────

export type ServerMessage =
  | { type: "auth.ok"; device_id: string; role: string; snapshot: Snapshot }
  | { type: "auth.error"; message: string }
  | { type: "pong" }
  | { type: "ack"; request_id: string; ok: true }
  | { type: "error"; request_id?: string; message: string }
  | { type: "state.tables"; tables: TableState[] }
  | { type: "state.order"; order: OrderState }
  | { type: "state.menu_item"; item_id: string; is_available: boolean }
  | { type: "state.kds_item"; item_id: string; status: string; order_id: string }
  | { type: "notify.item_ready"; order_id: string; item_name: string; table_number: string }
  | { type: "notify.ticket_done"; order_id: string; table_number: string }
  | { type: "payment.ok"; order_id: string; payment_id: string; change: number }
  | { type: "state.session"; session: ActiveSession }

// ── Shared state types ─────────────────────────────────────────────────────

export interface Snapshot {
  restaurant_id: string
  tables: TableState[]
  menu: MenuSnapshot
  active_session: ActiveSession | null
}

export interface TableState {
  id: string
  number: string
  label: string | null
  status: "free" | "occupied" | "bill-requested" | "reserved" | "blocked"
  floor_id: string
  capacity: number
  pos_x: number
  pos_y: number
  width: number
  height: number
  seating?: ActiveSeating
}

export interface ActiveSeating {
  id: string
  guests_count: number
  opened_at: string
  waiter_id: string | null
  order?: OrderState
}

export interface OrderState {
  id: string
  seating_id: string
  status: string
  subtotal: number
  discount_amount: number
  tax_amount: number
  total: number
  note: string | null
  items: OrderItemState[]
}

export interface OrderItemState {
  id: string
  menu_item_id: string
  menu_item_name: string
  qty: number
  unit_price: number
  discount: number
  modifiers: ModifierSelection[]
  course: number
  status: string
  note: string | null
  sent_at: string | null
  ready_at: string | null
}

export interface MenuSnapshot {
  categories: MenuCategory[]
  items: MenuItem[]
  groups: ModifierGroup[]
  modifiers: Modifier[]
}

export interface MenuCategory {
  id: string; name: string; sort: number; color: string | null; icon: string | null
}

export interface MenuItem {
  id: string; category_id: string; name: string; description: string | null
  photo_url: string | null; base_price: number; yield_g: number | null
  allergens: string[]; station: string; flags: string[]
  is_available: boolean; sort: number
}

export interface ModifierGroup {
  id: string; menu_item_id: string; name: string
  required: boolean; min_select: number; max_select: number; sort: number
}

export interface Modifier {
  id: string; group_id: string; name: string; price_delta: number; sort: number
}

export interface ActiveSession {
  id: string
  opened_at: string
  closed_at: string | null
  opening_cash: number
  total_sales: number
  total_cash: number
  total_card: number
  total_discount: number
  total_comps: number
  status: string
  pos_session_payins: { id: string; amount: number; note: string | null; created_at: string }[]
  pos_session_payouts: { id: string; amount: number; note: string | null; created_at: string }[]
}

// ── Helpers ────────────────────────────────────────────────────────────────

export function encode(msg: ServerMessage): string {
  return JSON.stringify(msg);
}

export function decode(raw: string): ClientMessage | null {
  try {
    return JSON.parse(raw) as ClientMessage;
  } catch {
    return null;
  }
}
